File: PA77.java
//Leetcode : 648. Replace Words
//Time Complexity: O(nk) , where n is the number of words max of (sentence , ddictionary) and k is the average length of word.
//Space Complexity; O(nk) ,where n is the number of words max of (sentence , ddictionary) and k is the average length of word.
class Solution {
    class TrieNode{
        String word;
        TrieNode[] children;
        TrieNode(){
            children= new TrieNode[26];
        }
    }
    
    TrieNode root = new TrieNode();
    public void buildTrie(String word){
        TrieNode n= root;
        for(int i=0;i<word.length();i++){
            char c= word.charAt(i);
            if(n.children[c-'a']==null){
                TrieNode t= new TrieNode();
                n.children[c-'a'] = t;
            }
            n= n.children[c-'a'];
        }
        n.word = word;
    }
    
    public String searchWord(String word){
        TrieNode n= root;
        for(int i=0;i<word.length();i++){
            char c= word.charAt(i);
            if(n.word!=null){
                return n.word;
            }
            if(n.children[c-'a']==null){
                return word;
            }
          n=n.children[c-'a'];
        }
        return word;
    }
    public String replaceWords(List<String> dict, String sentence) {
        if(sentence== null || sentence.length()==0) return sentence;
        for(String s: dict){
            buildTrie(s);
        }
        StringBuilder result= new StringBuilder();
        String[] words =sentence.split("\\s+"); 
        for(int i=0;i<words.length;i++){
            if(i==0){
                result.append(searchWord(words[i]));
            }
            else{
                result.append(" ");
                result.append(searchWord(words[i]));
                
            }
            
        }
        return result.toString();
        
    }
}

File: PA_77.py
#LC 648 - Replace Words
#Time Complexity - O(n.k) k being the average length of the word
#Space Complexity - O(n)
class TrieNode(object):
    
    def __init__(self):
        self.children = dict()
        self.endOfWord = False
        
class Solution(object):
    def replaceWords(self, dic, sentence):
        """
        :type dict: List[str]
        :type sentence: str
        :rtype: str
        """
        self.root = TrieNode()
        
        def insert(word):
            """
            Inserts a word into the trie.
            :type word: str
            :rtype: None
            """
            currNode = self.root
            for i in range(len(word)):
                char = word[i]
                if char not in currNode.children:
                    currNode.children[char] = TrieNode()
                currNode = currNode.children[char]
            currNode.endOfWord = True
            
        for i in dic:
            insert(i)
        
        ans = []
        sent = sentence.split(' ')
        for word in sent:
            currNode = self.root
            temp = ""
            for i in range(len(word)):
                if word[i] in currNode.children:
                    temp = temp + word[i]
                    currNode = currNode.children[word[i]]
                    if currNode.endOfWord:
                        break
                else:
                    ans.append(word)
                    break
            if currNode.endOfWord:
                ans.append(temp)  
        return(" ".join(ans))